gen3/bin/healthcheck.sh

```diff
@@ -0,0 +1,163 @@
+        shift
+        ;;
+      *)
+	echo "Unrecognized flag"
+	help
+	exit 1
+	;;
+    esac
+  done
+
+  # refer to k8s api docs for pod status info
+  # https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.13/#podstatus-v1-core
+  local allPods=$(g3kubectl get pods --all-namespaces -o json | \
+    jq -r '[
+      .items[] | {
+        name: .metadata.name, namespace: .metadata.namespace, phase: .status.phase, reason: .status.reason,
+        created: .metadata.creationTimestamp, waitingContainers: [ .status.containerStatuses[] | { state: .state, ready:.ready} ]
+      }
+    ]')
+
+  local evictedPods=$(echo $allPods | jq -r '[.[] | select(.reason == "Evicted") ]')
+  local pendingPods=$(echo $allPods | jq -r '[.[] | select(.phase == "Pending")]')
+  local unknownPods=$(echo $allPods | jq -r '[.[] | select(.phase == "Unknown")]')
+  local crashLoopPods=$(echo $allPods | jq -r '[.[] | select( .waitingContainers[].state.waiting.reason == "CrashLoopBackOff")]')
+  local terminatingTimeoutPods='[]'
+  local pendingTimeoutPods='[]'
+
+  # check for pods pending for more than 10 minutes
+  while read -r pod; do
+    local podDate=$(echo $pod | jq -r '.created')
+    local startTime=$(date --date="$podDate" '+%s')
+    local secsPassed=$(( $(date '+%s') - $startTime ))
+    if [[ $secsPassed -gt 600 ]]; then
+      pendingTimeoutPods=$(echo $pendingTimeoutPods | jq -r ". += [$pod]")
+    fi
+  done <<< "$(echo $pendingPods | jq -c '.[]')"
+
+  # check for terminating pods
+  # Unfortunately the only way to get termination duration is by grepping the `describe` output
+  while read -r pod; do
+    local statusLine=$(g3kubectl describe pod $(echo $pod | jq -r '.name') --namespace $(echo $pod | jq -r '.namespace') | grep "Status:" -m 1)
+    if [[ "$(echo $statusLine | awk '{ print $2 }')" == "Terminating" ]]; then
+      # check how long it's been terminating
+      local s=$(echo $statusLine | grep -oP "\d+(?=s)")
+      local m=$(echo $statusLine | grep -oP "\d+(?=m)")
+      local h=$(echo $statusLine | grep -oP "\d+(?=h)")
+      if [[ $m -gt 10 || $s -gt 300 || $h -gt 0 ]]; then
+        terminatingTimeoutPods=$(echo $terminatingTimeoutPods | jq -r ". += [$pod]")
+      fi
+    fi
+  done <<< "$(echo $allPods | jq -c '.[]')"
+
+  # check status of nodes
+  # https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.13/#nodestatus-v1-core
+  local allNodes=$(g3kubectl get nodes -o json | \
+    jq -r '[
+      .items[] | {
+        name: .metadata.labels."kubernetes.io/hostname", conditions: [ .status.conditions[] | { (.type): .status }] | add
+      }
+    ]')
+  local notReadyNodes=$(echo $allNodes | jq -r '[.[] | select(.conditions.Ready != "True")]')
+
+  # check internet access
+  local curlCmd="curl --max-time 15 -s -o /dev/null -I -w "%{http_code}" https://www.google.com"
+  if [[ $HOSTNAME == *"admin"* ]]; then # if in admin vm, run curl in fence pod
+    curlCmd="g3kubectl exec $(get_pod fence) -- $curlCmd"
```

Comment:
Use `gen3 pod fence` rather than `get_pod ...` - might want to get rid of that some day ...